server/src/main/java/org/elasticsearch/index/query/QueryRewriteInterceptor.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.query;
+
+import java.util.List;
+
+/**
+ * Enables modules and plugins to intercept and rewrite queries during the query rewrite phase on the coordinator node.
+ */
+public interface QueryRewriteInterceptor {
+
+    /**
+     * Returns a rewritten query if modifications are required; otherwise, returns the same provided {@link QueryBuilder} instance unchanged.
+     *
+     * @param context the {@link QueryRewriteContext} providing the context for the rewrite operation
+     * @param queryBuilder the original {@link QueryBuilder} to potentially rewrite
+     * @return the rewritten {@link QueryBuilder}, or the original instance if no rewrite was needed
+     */
+    QueryBuilder rewrite(QueryRewriteContext context, QueryBuilder queryBuilder);
+
+    static QueryRewriteInterceptor multi(List<QueryRewriteInterceptor> interceptors) {
+        if (interceptors.isEmpty()) {
+            return (c, q) -> q;
+        }
+        return new Multi(interceptors);
+    }
```

Comment:
This implies some sort of priority. Meaning, the first rewriter could mutate queries in such a way that subsequent rewriters will never do a rewrite. 

I don't think we should ever allow more than one rewriter, otherwise we are at the whims of the order in which the plugins are loaded. 

Also, I am thinking we should be strict about this and only allow our plugin to do this.

Reply:
By doing this, we are allowing any ES plugin to add code like this (even ones we don't author). Meaning, there are many chances here for things to be abused or cause conflicts with eachother. 

How would we ever debug a custom plugin that overwrites this, returns something called "OurSpecialMatchQueryBuilder" before it gets passed to our rewriter? We would just never do the right thing and it would break.

Reply:
The idea here is that we will also use this type of rewriting for knn and sparse vector fields. So while we could put all of that in a single class, I thought that having specific interceptors based on the query type we're intercepting may be easier to read. 

As far as the issue of overriding it, maybe we can resolve that with documentation? There are lots of ways that users can break things with custom plugins, but that's not a supported path and they shouldn't do that, no?

Reply:
Yes this is not ideal. The intent here is use internally so if we have a way to restrict this from being used outside that would be great. Maybe it shouldn't be a plugin and we just allow xpack to register this type of rewriter?

Reply:
OK. Thinking about this a little more, what about this as a compromise? 

Instead of a List of `QueryRewriteInterceptor` we can define a Map keyed on the name of the query (e.g. `match`). If multiple interceptors are found for a single type of query, we can fail fast. 

I think it's OK to allow plugins to create their own interceptors, because there could be valid reasons to do this, but it alleviates the issue of priority and order of operations at least. Would this be an acceptable compromise?

Reply:
That sounds like a good compromise to me.

Reply:
> we can define a Map keyed on the name of the query (e.g. match). If multiple interceptors are found for a single type of query, we can fail fast.

failing fast if there is more than one seems OK. Let's just be sure to log the plugin names that are causing the failure and indicate that only one of them can be loaded.

Reply:
Updated in 9640ad11d5b437d8962374ecb23cde4537667139

Reply:
Performed module refactoring in c998f1cc6c433881c2a270a235d8e97ae3f6495d to alleviate security concerns.